[Woo POS][Products search] Make remote products search more predictable

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-149
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR improves the loading state for searches, and debounces each search preventing multiple requests being made for intermediate keystrokes.

As part of this PR, a previously required workaround for pull-to-refresh issues was removed.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Enable the product search feature flag
2. Launch the app and navigate to POS
3. Search for products – observe (using a network proxy tool) that searches are only sent after a break in typing (apart from the first letter)

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

The current behaviour is to start a search immediately when a key is typed, so that we show the loading state. Further keypresses are debounced by 300ms, reducing the number of searches while the user is typing.

We have to keep track of whether a search finished, to ensure that we correctly debounce new searches.

This behaviour might change with the development of the final UI – e.g. it may feel better to keep the empty state on screen while the initial search term is typed out, so not show the loading indicator immediately.

At the moment, we also reuse the same ItemView rather than layering a new one on top. That doesn't have any functionality implications at this stage, but with the new component and empty state it will probably not remain that way much longer.

I've tested this on an iPad Air running iOS 17.7.3

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/f17bb3f7-6d01-472a-9931-1b8b953808dc


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are adde

joshheald committed:
Ensure the search list doesn’t show fullscreen load

joshheald committed:
Only use the minimal loading state for searching

joshheald committed:
Stop changing the state when requests are cancelled

Changing state when requests are cancelled was used as a workaround for refreshing issues which is no longer required due to correct handling of the tasks.

The workaround caused issues with continuously showing the search state while changing the input, especially when debounced.

joshheald committed:
Debounce and cancel intermediate searches

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
Ensure we consistently show the loading state

The current behaviour is to start a search immediately when a key is typed, so that we show the loading state. Further keypresses are debounced by 300ms, reducing the number of searches while the user is typing.

We have to keep track of whether a search finished, to  ensure that we correctly debounce new searches.

joshheald committed:
Merge branch 'trunk' into woomob-149-make-search-more-predictable

staskus reviewed (approved):
Works well. The debounce implementation itself wasn't that easy to read in the beginning but going line by line I understand why it was done this way. We have `Combine` debounce but it wouldn't quite give the same result since we wouldn't disable debouncing when the request finishes.

joshheald commented:
@staskus thanks for the review

> The debounce implementation itself wasn't that easy to read in the beginning but going line by line I understand why it was done this way. 

I'll see if I can make it any clearer...

> We have Combine debounce but it wouldn't quite give the same result since we wouldn't disable debouncing when the request finishes.

Yeah. I'm trying to avoid Combine in the view layer anyway, since `Observation` seems to be Apple's direction of travel and we can't use any publisher chains on things we get from the aggregate model. I don't think we could use it on an `@State` either.

joshheald commented:
The other option would be not to use the `loading` state for search, and to keep `isSearching` as view state, same as we do for Pull to Refresh. `loading` would then only be used for pagination, which is fine. I'll think about this in future PRs.

joshheald committed:
Explain debouncing logic

joshheald committed:
Merge branch 'trunk' into woomob-149-make-search-more-predictable
